_posts/2019-01-30-year-of-the-dog.md

```diff
@@ -0,0 +1,199 @@
+   logic had to be implemented in the frontend** and merely validated by the
+   API.
+2. **State management and synchronization in the React apps was getting too
+   complex** - some operations and queries required several network calls and
+   dispatching multiple redux actions. Some critical operations were not
+   atomic, when they should have been.
+3. **Onboarding new developers to our redux-saga based abstractions** and all
+   the related conventions **proved challenging**, as documentation was not as
+   comprehensive as it should have been and old conventions coexisted with new
+   ones.
+4. With most of the data in our apps coming from the API, we felt that manually
+   writing Typescript interfaces for all these resources was a lot of error
+   prone manual work. We thought about generating them from our
+   [OpenAPI/Swagger](https://swagger.io/specification/) documentation, but the
+   experiment wasn't very conclusive. Coupled with the type information loss
+   from the redux-saga generators, we felt that we were **not realizing
+   Typescript's potential**.
+5. **Documentation was maintained manually as a Swagger specification**. This
+   was also a lot of error-prone, manual work.
+
+Faced with these problems, we decided to try and address them by progressively
+introducing a new GraphQL API, in the same Rails codebase as the existing
+RESTful API. Here is what we observed:
+
+1. The additional layer gave us more room and structure to **move business
+   logic back to the backend**.
+2. **Data fetching in the frontend is now a lot less complex**, with queries
+   mapping more or less 1-to-1 with views, and mutations to forms and buttons.
+   We are not using any redux-based abstraction for CRUD anymore, and the
+   overall amount of state management code has decreased. State is (or
+   appears) also more localized to where it belongs in the component tree. We
+   found this easier to reason about.
+3. The simplicity, but also the adherence to a well-adopted and documented
+   standard for data fetching made **onboarding easier**. There are a lot of
+   good GraphQL resources for frontend developers. In our experience picking up
+   the technology was easy for everyone.
+4. The tooling is amazing, it lets us **leverage Typescript's strengths by
+   generating precise types for our API data**. We use apollo-client and
+   apollo-codegen, they are well-documented, worked fine and the caching layer
+   is the cherry on top.
+5. **Our API documentation is now in much better shape**, because most of it is
+   automated, keeping it in sync with the implementation is trivial, and
+   [GraphiQL][graphiql] offers a good interface to explore it.
+
+It took some time to have a viable minimum implementation, but once this was
+deployed, it became clear very quickly that we wanted to go in that direction.
+
+Two more pain-points that we did not identify before migrating but are now obvious:
```

Comment:
I would rephrase this a little bit. For me it read like we had problems but it's actually about `pain-points that were solved`